### PR TITLE
[crypto & al] Direct signing/verification for values

### PR DESCRIPTION
**Pull request by huitseeker**
## What this does

We used to do:
```
let hash = msg.hash()
private_key.sign_message(hash)
signature.verify_msg(hash, public_key)
public_key.verify_msg(hash, signature)
```

Now we do:
```
private_key.sign(&message)
signature.verify(&message, public_key)
public_key.verify(&message, signature)
```
Throughout the code base.

### Why was it this way?

Long story short: domain separation & weaknesses in serialization engine. We bypassed serialization (which was ambiguous, as of a *while ago*) and had a very specific way of constructing a hash value, which was `sha3(sha3(domain_separation_bytes) || serialized_object_bytes)`, and then you signed that `HashValue`. See `libra_crypto::hash::CryptoHasher` for the domain separation details.

### What are the signed bytes now?

We automatically derive them for you, but they're `sha3(domain_separation_bytes) || serialized_object_bytes`.

### So signing is handled for existing `struct`s. What do I do if I want to define `MyFancyStruct` ?

```
use libra_crypto_derive::{CryptoHasher, LCSCryptoHash}

#[derive(Serialize, Deserialize, CryptoHasher, LCSCryptoHash)]
pub struct MyCryptoSignableCustomThingy {
...
}

priv_key.sign(my_crypto_signable_custom_thingy)
```

### Is this in specs?

It's upcoming. The older signing/verification isn't there yet.

## Why is this better?

It's simpler and more secure (you might pass the wrong hash in code, it's even a [crypto gotcha](https://github.com/SalusaSecondus/CryptoGotchas#signatures)).

## Why is this worse?

Serialization is now on the critical path for signing / verification operations, which are exposed as infallible. Serialization can fail in some edge cases, so now they can panic. There's no clean way of recovering from this sort situation, but we are committed to and have a plan to push serialization failures into a never-visited corner (see e.g. #4928).

## What's left?

See above re:spec.  It's my very next item.

### CODE REVIEW

File: testsuite/generate-format/src/consensus.rs [hunk 1]
```diff
@@ -3,30 +3,33 @@
...
 use rand::{rngs::StdRng, SeedableRng};
+use serde::{Deserialize, Serialize};
 use serde_reflection::{Registry, Result, Samples, Tracer, TracerConfig};
 
 /// Return a relative path to start tracking changes in commits.
 pub fn output_file() -> Option<&'static str> {
     Some("tests/staged/consensus.yaml")
 }
 
+/// This aims at signing canonically serializable LCS data
```
Comment (matbd) [inline testsuite/generate-format/src/consensus.rs:21]: LGTM

File: testsuite/generate-format/src/libra.rs [hunk 2]
```diff
@@ -3,31 +3,41 @@
...
 
+    let message = TestLibraCrypto("Hello, World".to_string());
+
     let mut rng: StdRng = SeedableRng::from_seed([0; 32]);
     let private_key = Ed25519PrivateKey::generate(&mut rng);
     let public_key: Ed25519PublicKey = (&private_key).into();
-    let signature = private_key.sign_message(&hashed_message);
+    let signature = private_key.sign(&message);
 
+    // This is just a means to an end to record the custom de/ser of hashes & signatures
```
Comment (matbd) [inline testsuite/generate-format/src/libra.rs:39]: Remove redundant comment?

File: crypto/crypto/src/unit_tests/ed25519_test.rs [hunk 3]
```diff
@@ -161,7 +161,7 @@ proptest! {
         keypair in uniform_keypair_strategy::<Ed25519PrivateKey, Ed25519PublicKey>()
     ) {
         let hashable = CryptoHashable(x);
-        let signature = keypair.private_key.sign(&hashable).expect("all `RawTransaction` objects should LCS-serialize correctly");
```
Comment (matbd) [inline crypto/crypto/src/unit_tests/ed25519_test.rs:164]: So, this was the only occurrence of the faillible `sign`? Everybody is hashing first then using sign_message, I suppose?

File: secure/storage/src/vault.rs [hunk 4]
```diff
@@ -280,22 +281,34 @@ impl CryptoStorage for VaultStorage {
...
+        name: &str,
+        message: &T,
+    ) -> Result<Ed25519Signature, Error> {
         let name = self.crypto_name(name);
-        Ok(self.client.sign_ed25519(&name, message.as_ref(), None)?)
+        let mut bytes = <T::Hasher as libra_crypto::hash::CryptoHasher>::seed().to_vec();
+        lcs::serialize_into(&mut bytes, &message)
+            .map_err(|_| libra_crypto::traits::CryptoMaterialError::SerializationError)
+            .expect("Serialization of signable material should not fail.");
+        // TODO: triple-check this output is valid
```
Comment (matbd) [inline secure/storage/src/vault.rs:294]: Is this TODO necessary?

File: crypto/crypto/src/unit_tests/ed25519_test.rs (hunk 3 above)
Comment (huitseeker) [reply crypto/crypto/src/unit_tests/ed25519_test.rs:164]: All the `sign` instances added herein are still faillible.

>  Everybody is hashing first then using sign_message, I suppose?

Yes.

File: crypto/crypto/src/traits.rs [hunk 5]
```diff
@@ -121,20 +121,13 @@ pub trait SigningKey:
...
-    /// For the moment, this signature is incompatible with the conversion into
-    /// a `HashValue` above. We intend to deprecate `sign message` in favor of
-    /// the present function soon.
-    fn sign<T: CryptoHash + Serialize>(
-        &self,
-        message: &T,
-    ) -> Result<Self::SignatureMaterial, CryptoMaterialError>;
+    /// Note: this assumes serialization is unfaillible. See libra_common::lcs::ser
+    /// for a discussion of this assumption.
+    fn sign<T: CryptoHash + Serialize>(&self, message: &T) -> Self::SignatureMaterial;
```
Comment (matbd) [inline crypto/crypto/src/traits.rs:130]: Yay!

**Comment by ankushagarwal** [comment]
@huitseeker Can you update the swiss-knife [README.md](https://github.com/libra/libra/blob/master/client/swiss-knife/README.md) before you land?

File: client/swiss-knife/README.md [hunk 6]
```diff
@@ -98,7 +99,7 @@ $ cargo run -p swiss-knife -- generate-test-ed25519-keypair --seed 0
 ## Sign a payload using a Ed25519 Private Key
```
Comment (ankushagarwal) [inline client/swiss-knife/README.md:99]: nit: Sign a raw_txn using a Ed25519 Private Key

File: client/swiss-knife/README.md [hunk 7]
```diff
@@ -20,6 +20,7 @@ The `sample_inputs` folder contains a list of sample json inputs for various ope
 # For the txn_params json schema, look at the `struct TxnParams`
 # For the script_params json schema, look at the `struct MoveScriptParams`
 $ cargo run -p swiss-knife -- generate-raw-txn < sample_inputs/generate_raw_txn_peer_to_peer_transfer.json
+
```
Comment (ankushagarwal) [inline client/swiss-knife/README.md:23]: nit: Remove `raw_txn_hash` from line 28.

**Pull request closed without merging**